llvm/tools/llvm-dwarfdump/Coverage.cpp

```diff
@@ -0,0 +1,257 @@
+//===-- Coverage.cpp - Debug info coverage metrics ------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm-dwarfdump.h"
+#include "llvm/BinaryFormat/Dwarf.h"
+#include "llvm/DebugInfo/DIContext.h"
+#include "llvm/DebugInfo/DWARF/DWARFAcceleratorTable.h"
+#include "llvm/DebugInfo/DWARF/DWARFCompileUnit.h"
+#include "llvm/DebugInfo/DWARF/DWARFContext.h"
+#include "llvm/IR/CFG.h"
+#include "llvm/IR/DebugInfoMetadata.h"
+#include "llvm/IR/DebugProgramInstruction.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IRReader/IRReader.h"
+#include "llvm/Object/ObjectFile.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/SourceMgr.h"
+#include <set>
+
+using namespace llvm;
+using namespace llvm::dwarf;
+using namespace llvm::object;
+
+typedef std::pair<std::string, std::string> StringPair;
+/// Pair of file index and line number representing a source location.
+typedef std::pair<uint16_t, std::size_t> SourceLocation;
+
+/// Returns the set of source lines covered by a variable's debug information.
+static std::optional<std::set<SourceLocation>>
```

Comment:
Recommend replacing uses of `std::set` here with [DenseSet](https://llvm.org/docs/ProgrammersManual.html#llvm-adt-denseset-h), which should be more efficient for this case (see also [here](https://llvm.org/docs/ProgrammersManual.html#set) - the advantages of std::set aren't used here). For the set intersection, there is `llvm::set_intersect` (needs `#include "llvm/ADT/SetOperations.h"`) which will work for DenseSet.